test/functional/p2p_leak.py

```diff
@@ -106,39 +94,45 @@ def set_test_params(self):
         self.num_nodes = 1
 
     def run_test(self):
-        no_version_disconnect_node = self.nodes[0].add_p2p_connection(
-            CNodeNoVersionMisbehavior(), send_version=False, wait_for_verack=False)
-        no_version_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVersionIdle(), send_version=False, wait_for_verack=False)
-        no_verack_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVerackIdle(), wait_for_verack=False)
+        # Peer that never sends a version. We will send a bunch of messages
+        # from this peer anyway and verify eventual disconnection.
+        no_version_disconnect_peer = self.nodes[0].add_p2p_connection(
+            LazyPeer(), send_version=False, wait_for_verack=False)
+
+        # Another peer that never sends a version, nor any other messages. It shouldn't receive anything from the node.
+        no_version_idle_peer = self.nodes[0].add_p2p_connection(LazyPeer(), send_version=False, wait_for_verack=False)
+
+        # Peer that sends a version but not a verack.
+        no_verack_idle_peer = self.nodes[0].add_p2p_connection(NoVerackIdlePeer(), wait_for_verack=False)
 
-        # Send enough veracks without a message to reach the peer discouragement
-        # threshold. This should get us disconnected.
+        # Send enough ping messages (any non-version message will do) prior to sending
+        # version to reach the peer discouragement threshold. This should get us disconnected.
         for _ in range(DISCOURAGEMENT_THRESHOLD):
-            no_version_disconnect_node.send_message(msg_verack())
+            no_version_disconnect_peer.send_message(msg_ping())
 
-        # Wait until we got the verack in response to the version. Though, don't wait for the other node to receive the
+        # Wait until we got the verack in response to the version. Though, don't wait for the node to receive the
         # verack, since we never sent one
-        no_verack_idlenode.wait_for_verack()
+        no_verack_idle_peer.wait_for_verack()
 
-        wait_until(lambda: no_version_disconnect_node.ever_connected, timeout=10, lock=mininode_lock)
-        wait_until(lambda: no_version_idlenode.ever_connected, timeout=10, lock=mininode_lock)
-        wait_until(lambda: no_verack_idlenode.version_received, timeout=10, lock=mininode_lock)
+        wait_until(lambda: no_version_disconnect_peer.ever_connected, timeout=10, lock=mininode_lock)
```

Comment:
oops. Something seems to have gone wrong with the rebase here. You've changed these lines in the first commit (_test: remove `CNodeNoVersionMisbehavior` from p2p_leak.py _), but the names are only changed above in the third commit (_test: Distinguish between nodes(bitcoind) and peers(mininodes) in p2p_leak.py_), which means that the test fails for the first two commits. We try to make all of our commits clean so that tools like git bisect work.

Reply:
Thank you, @jnewbery. I learnt to:
1. Not rebase latest master every round of review
2. Make sure tests pass on every commit

I have pushed up fixed commits.